Specular Energy Conservation fix

MiiBond opened the pull request:
Okay, I was starting to work on supporting the OpenPBR specular model (coloured specular right out to F90 and then, after that, texturable IOR and the F82 model for conductors).  However, right away, I noticed that the current diffuse/specular logic isn't correct for dielectrics. We multiply the diffuse part by F0, leaving too much energy at glancing angles.

Here's a furnace test where metallic increases from left to right and metallicF0Factor increases from bottom to top.
![image](https://github.com/user-attachments/assets/c7dabe4f-addc-41e5-a918-fe3f261b4d61)

This PR eliminates this completely.

I've only touched the metallic workflow (I think) as this is all that's applicable to OpenPBR compatibility (and that's my current focus). Before I continue further, I'd like to hear some feedback as I've made some other changes.
1. I'm passing the full reflectivity colour into the shader now rather than pre-multiplying by F0. This gives us more control in the shader and will be necessary later for coloured F90 and texturable IOR.
2. In `pbrBlockFinalLitComponents`, I'm multiplying the irradiance by 1.0 - reflectance to account for reflected energy that doesn't penetrate the surface.
[https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge#KQYNYS#10](https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge#KQYNYS#10)

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

MiiBond commented:
Outside of a furnace test, the change can be pretty subtle but you can see the difference around edges:
Before:
![image](https://github.com/user-attachments/assets/75289827-8744-4e25-b456-550842cfbf57)
After:
![image](https://github.com/user-attachments/assets/6f476912-1e11-41df-ae56-57985bda8f09)

bjsplat commented:
Snapshot stored with reference name:
refs/pull/16476/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/16476/merge
https://gui.babylonjs.com/?snapshot=refs/pull/16476/merge
https://nme.babylonjs.com/?snapshot=refs/pull/16476/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge#BCU1XR#0

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -117,17 +119,21 @@ reflectivityOutParams reflectivityBlock(
             #endif
 
             // Compute the converted diffuse.
-            outParams.surfaceAlbedo = mix(baseColor.rgb * (1.0 - metallicF0), vec3(0., 0., 0.), metallicRoughness.r);
```
This `(1.0 - metallicF0)` factor isn't correct. It compensates for energy lost to specular at F0 but ignores energy lost at other angles.

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockFinalLitComponents.fx:
```diff
@@ -24,6 +24,11 @@ aggShadow = aggShadow / numLights;
 
+    #ifdef METALLICWORKFLOW
+        // Account for energy loss due to specular reflectance
+        finalIrradiance *= (1.0 - subSurfaceOut.specularEnvironmentReflectance);
```
This is the main change. The irradiance needs to be scaled by `1 - specular reflectance`

sebavan replied:
Why is this not always needed ?

sebavan replied:
This was probably the case too in the previous IOR PR ?

MiiBond replied:
Yes, this would be needed for the spec-gloss workflow too. I just wanted to affect as little code as possible as there are a LOT of parameters and code paths in the PBRMaterial...
I'll take a stab at fixing the spec-gloss flow as well.

MiiBond replied:
To get the furnace test to pass for the spec-gloss workflow, I mainly have to replace `reflectance` with `specularEnvironmentReflectance` in the energy compensation (`(1.-reflectance)` for diffuse).
However, there's also an `energyConservationFactor` that's multiplied by the radiance and this is causing issues with energy conservation. Removing it allows the furnace test to pass perfectly.
Any ideas what this code is actually supposed to be doing? It seems strange to multiply radiance by a conservation factor and not do anything to irradiance. Is this to compensate for energy lost on rough surfaces? If needed, perhaps it should be incorporated to the `specularEnvironmentReflectance` so that any energy added to the specular lobe is subtracted from the diffuse lobe.

MiiBond replied:
Making these changes results in a perfect furnace test but lots and lots of changes in the visualization tests :(

As an example, [this playground](https://playground.babylonjs.com/#Z1VL3V#4) shows the change with coloured specular. Before, the F0 was being taken as the maximum component of the colour which, in this case, was 1.0. Now, however, F0 is the specular colour and the diffuse has this energy removed. This is more correct, I believe, but it's a big change. Dielectrics with coloured specular are pretty rare in the real world but I'm not sure about Babylon users.
<img width="360" alt="image" src="https://github.com/user-attachments/assets/7549bd68-f4d0-476e-836c-935ed5016864" />
<img width="673" alt="image" src="https://github.com/user-attachments/assets/a1fe2eb7-02c3-43dd-aa84-42bf4a9125a6" />

MiiBond replied:
Perhaps I should add these changes to a separate PR to give us more time to digest the impact of the spec-gloss workflow change?

Popov72 replied:
cc @sebavan, as I'm not able to answer your questions.

There's also [this PR](https://playground.babylonjs.com/#8F5HYV#73) which also looks quite different (and use the metallic workflow).

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

bjsplat commented:
Snapshot stored with reference name:
refs/pull/16476/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/16476/merge
https://gui.babylonjs.com/?snapshot=refs/pull/16476/merge
https://nme.babylonjs.com/?snapshot=refs/pull/16476/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge#BCU1XR#0

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

sebavan reviewed (commented):
I love it and will merge after @Popov72 review tomorrow !!! Thanks a lot for the fix !!!

MiiBond commented:
I've reverted the renaming of `vReflectivityColor` and `vMetallicReflectanceFactors` because it was causing a headache with NodeMaterials...
I don't like the existing naming but it really isn't necessary to change that for this PR.

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

MiiBond commented:
> We had a bug with refraction transmittance, where we had to change the factor to 1.0 - max(specularEnvironmentReflectance.r,

Do you remember any more details about the bug? On the face of it, I can't see an issue. If specular reflectance is coloured, that should also affect the light that refracts through the surface.

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

sebavan commented:
@Popov72 would be great to find back the gltf model to test it ?

MiiBond commented:
> See this post in our forum

Ah, okay. This PR doesn't touch the subsurface block so we're good.
And I tested that asset and it looks identical.

sebavan commented:
@MiiBond only one comment and it will go in 
![image](https://github.com/user-attachments/assets/dada3fc7-81ac-400a-8d89-171d983ccc60)

MiiBond commented:
Yes, it would apply in both workflows. I just didn't want to touch more than I needed to since this change was already affecting a lot of visualization tests.

sebavan commented:
@MiiBond Can you check in a subsequent PR to remove the METALLICWORKFLOW conditions ? and I merge this one ?

MiiBond commented:
> Can you check in a subsequent PR to remove the METALLICWORKFLOW conditions ? and I merge this one ?

Yes, I think that would be best.

MiiBond commented:
Actually, please hold off on merging this. I want to make some changes.

MiiBond commented:
Okay, I've made some small but significant changes.
glTF and OpenPBR both define the energy that passes through the specular layer as being greyscale, even when the specular colour is not. This is for the benefit of artists so they don't have to deal with their material's diffuse colour changing in the opposite tinting of the specular.

Specular transmission in glTF:
```
max_value(fresnel(ior, specular_color))
```

Specular transmission in OpenPBR:
```
fresnel(ior)
```

For non-white specular colours, this DOES result in energy lost but the same is true for non-white albedo so I suppose this can be thought of as the specular layer absorbing the energy. Anyway, energy isn't gained and the furnace test passes for white albedo and white specular colour.

Later, we can add different specular reflectance modes but the current PR should be maximally backwards-compatible while fixing the basic energy conservation problem at F90.

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

virtualzavie commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockFinalLitComponents.fx:
```diff
@@ -24,6 +24,18 @@ aggShadow = aggShadow / numLights;
     vec3 finalIrradiance = reflectionOut.environmentIrradiance;
 
+    #if defined(METALLICWORKFLOW) || defined(SPECULAR_GLOSSINESS_ENERGY_CONSERVATION)
+        // Account for energy loss due to specular reflectance
+        // glTF specifies that diffuse energy is reduced by max_value(fresnel(ior, specular_color))
+        vec3 specEnergy = vec3(max(subSurfaceOut.specularEnvironmentReflectance.r, max(subSurfaceOut.specularEnvironmentReflectance.g, subSurfaceOut.specularEnvironmentReflectance.b)));
```
👍

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/lightFragment.fx:
```diff
@@ -104,7 +104,11 @@
+                        #ifdef METALLICWORKFLOW
+                            info.specular = computeSpecularLighting(preInfo, normalW, clearcoatOut.specularEnvironmentR0, specularEnvironmentR90, AARoughnessFactors.x, diffuse{X}.rgb, vReflectivityColor.b);
+                        #else
+                            info.specular = computeSpecularLighting(preInfo, normalW, clearcoatOut.specularEnvironmentR0, specularEnvironmentR90, AARoughnessFactors.x, diffuse{X}.rgb, 1.5);
```
It's the IOR, yeah. Here, we pass in the IOR to artificially scale the reflectance so that IOR's near 1.0 have zero reflectance.
The IOR is only available in the metallic workflow so I'm just passing the default 1.5 in other cases.

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

MiiBond commented:
One of the things that I've been trying to do is to separate out the dielectric and metallic lobes and then blend them together at the end based on the metallic value. This makes for clearer code and fewer potential mistakes in blending lobes. Unfortunately, this affects a lot of other logic. For example, iridescence expects a single `specularEnvironmentR0` value that is already a blend between dielectric and metallic.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

MiiBond commented:
Okay, I've pushed more changes to clean things up and apply the same logic to spec-gloss and metal-rough.
In the process of doing this, I've noticed that the logic of lerping baseColor to black for metals results in double energy loss for the dielectric layer. This means that we lose too much energy for materials with metallic values in between 0 and 1. I know that this is the same logic as the code that the glTF appendix B gives us:
```
const black = 0

c_diff = lerp(baseColor.rgb, black, metallic)
f0 = lerp(0.04, baseColor.rgb, metallic)
α = roughness^2

F = f0 + (1 - f0) * (1 - abs(VdotH))^5

f_diffuse = (1 - F) * (1 / π) * c_diff
f_specular = F * D(α) * G(α) / (4 * abs(VdotN) * abs(LdotN))

material = f_diffuse + f_specular
```

However, it's wrong for these in-between metallic values. The Fresnel value should be enough to properly blend between dielectric and metal and that's what I've set out to do. It required quite a few changes.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

MiiBond commented:
Another example of a fix that falls out of these changes:
<img width="389" alt="transparent transitions" src="https://github.com/user-attachments/assets/e66d0368-0ea1-4e87-97ab-ca237e10d6f7" />

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

Popov72 reviewed (commented):
I really like your changes (there are still a few things I don't understand, if you could clarify them in the comments), but as you pointed out, many visualization tests are now failing, and we know that these changes will break some of our partners' code...

I'm not sure if we should add another flag to maintain backward compatibility or do something else... cc @sebavan

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -117,17 +121,29 @@ reflectivityOutParams reflectivityBlock(
+            // It represents the total percentage of light that is reflected at normal incidence.
+            // In glTF's material model, this is the F0 value calculated from the IOR and then multiplied by the maximum component of the specular colour.
+            outParams.reflectanceF0 = mix(reflectivityColor.a * metallicReflectanceFactors.a, 1.0, outParams.metallic);
+            outParams.reflectanceF90 = vec3(mix(metallicReflectanceFactors.a, 1.0, outParams.metallic));
```
Oh, good catch. Thank you. I love being able to clean some of this up.
On that topic, could we just remove the `FROSTBITE_REFLECTANCE` code? It actually looks to me like it wouldn't even compile anymore if someone defined that flag.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockFinalLitComponents.fx:
```diff
@@ -24,6 +25,21 @@ aggShadow = aggShadow / numLights;
     vec3 finalIrradiance = reflectionOut.environmentIrradiance;
 
+    #if defined(METALLICWORKFLOW) || defined(SPECULAR_GLOSSINESS_ENERGY_CONSERVATION)
+        // Account for energy loss due to specular reflectance
+        vec3 baseSpecularEnergy = vec3(baseSpecularEnvironmentReflectance);
+        #if defined(ENVIRONMENTBRDF)
+            #ifdef MS_BRDF_ENERGY_CONSERVATION
+                baseSpecularEnergy *= baseSpecularEnergyConservationFactor;
+            #endif
+        #endif
+        finalIrradiance *= (vec3(1.0) - baseSpecularEnergy);
+        #if defined(SS_REFRACTION)
+            // Decrease transmission in the same way as the irradiance
+            subSurfaceOut.finalRefraction *= (vec3(1.0) - baseSpecularEnergy);
+        #endif
```
Any light energy that is reflected by the specular layer must not impact the diffuse layer. So, the conservation factor (that brightens rough reflections, especially for metals) used on the specular layer needs to also be subtracted from the diffuse layer.
Same deal with the transmitted light (`finalRefraction`). I've move the energy compensation logic here so that we have access to things like the conservation factor that we wouldn't otherwise have.

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockFinalUnlitComponents.fx:
```diff
@@ -1,6 +1,15 @@
 // _____________________________ Diffuse ________________________________________
 vec3 finalDiffuse = diffuseBase;
+#ifndef UNLIT
+    #if defined(ENVIRONMENTBRDF) && defined(MS_BRDF_ENERGY_CONSERVATION)
+        finalDiffuse *= (1.0 - baseSpecularEnvironmentReflectance * baseSpecularEnergyConservationFactor) * surfaceAlbedo;
+    #else
+        finalDiffuse *= (1.0 - baseSpecularEnvironmentReflectance) * surfaceAlbedo;
+    #endif
```
You're right. I should't be using the IBL values here. I think I was trying to compensate for my removal of the `mix(surfaceAlbedo, black, metallic)` from the reflectivity code and I didn't bother looking into the analytical light code enough.
After actually looking at this code, I see that the correct solution would be to refactor `computeDiffuseLighting` and `computeSpecularLighting` to handle the `(1 - F) ` logic for the diffuse lobe. Maybe a separate function for computing the Fresnel so that this value could be passed into both functions? Perhaps a separate PR for that? I think the effect will be fairly minor.

MiiBond replied:
Incidentally, why is `diffuseFinal` calculated inside of `pbrBlockFinalUnlitComponents.fx` and `specularFinal`, `irradianceFinal`, etc. are all calculated in `pbrBlockFinalLitComponents.fx`?

MiiBond replied:
Also, I'm just looking more closely at the Fresnel calculation and we're using VdotH in `fresnelSchlickGGX`. This seems incorrect to me as the half vector is between V and L and isn't involving N at all. The cosθ term in the Schlick approximation is the cosine of half the angle between incoming and outgoing light directions. It has to involve the surface normal because that's how the outgoing light direction is calculated.
Indeed, if I output the result of `fresnelSchlickGGX` in the shader, it's a uniform value across the geometry which definitely seems wrong.
But perhaps I'm missing something?

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -103,7 +107,7 @@ reflectivityOutParams reflectivityBlock(
         // Diffuse is used as the base of the reflectivity.
         vec3 baseColor = surfaceAlbedo;
-
+        outParams.metallic = metallicRoughness.r;
```
Done

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -117,17 +121,29 @@ reflectivityOutParams reflectivityBlock(
-
             #if DEBUGMODE > 0
-                outParams.metallicF0 = metallicF0;
+                outParams.metallicF0 = vec3(reflectivityColor.a) * metallicReflectanceFactors.rgb;
```
Good catch. Done.

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -117,17 +121,29 @@ reflectivityOutParams reflectivityBlock(
+            surfaceReflectivityColor = metallicReflectanceFactors.rgb;
+
+            // Final F0 for dielectrics = F0 * specular_color * specular_weight
+            vec3 dielectricColorF0 = vec3(reflectivityColor.a * metallicReflectanceFactors.rgb * metallicReflectanceFactors.a);
```
Done

bjsplat commented:
Building or testing the sandbox has failed.

If the tests failed, results can be found here:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/testResults/

bjsplat commented:
Building or testing the playground has failed.

If the tests failed, results can be found here:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/testResults/

bjsplat commented:
Snapshot stored with reference name:
refs/pull/16476/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/16476/merge
https://gui.babylonjs.com/?snapshot=refs/pull/16476/merge
https://nme.babylonjs.com/?snapshot=refs/pull/16476/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/16476/merge#BCU1XR#0

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have changed file(s) that made possible changes to the sandbox.
You can test the sandbox snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/SANDBOX/refs/pull/16476/merge/

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have made possible changes to the playground.
You can test the snapshot here:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/

The snapshot playground with the CDN snapshot (only when available):

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/PLAYGROUND/refs/pull/16476/merge/?snapshot=refs/pull/16476/merge

Note that neither Babylon scenes nor textures are uploaded to the snapshot directory, so some playgrounds won't work correctly.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

sebavan commented:
Moving to draft until we discuss the back compat more to prevent too early integration

MiiBond commented:
Okay, I've added a flag to the `PBRBRDFConfiguration` to enable the legacy behaviour. It's called `useLegacySpecularEnergyConservation`.

I also added the two furnace test visualizations for metal-rough and spec-gloss.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -123,17 +121,40 @@ reflectivityOutParams reflectivityBlock(
-            // Compute the converted reflectivity.
-            surfaceReflectivityColor = mix(metallicF0, baseColor, metallicRoughness.r);
+            #ifdef LEGACY_SPECULAR_ENERGY_CONSERVATION
+                outParams.surfaceAlbedo = baseColor.rgb * (1.0 - dielectricF0) * (1.0 - outParams.metallic);
```
```suggestion
                outParams.surfaceAlbedo = baseColor.rgb * (vec3(1.0) - vec3(dielectricF0) * surfaceReflectivityColor) * (1.0 - outParams.metallic);
```
because `metallicF0 = vec3(dielectricF0) * surfaceReflectivityColor`

Popov72 commented on packages/dev/core/src/Materials/Node/Blocks/PBR/subSurfaceBlock.ts:
```diff
@@ -188,7 +188,7 @@ export class SubSurfaceBlock extends NodeMaterialBlock {
                 , vTintColor
                 , normalW
-                , specularEnvironmentReflectance
+                , cumulativeSpecularEnvironmentReflectance
```
```suggestion
        #ifdef LEGACY_SPECULAR_ENERGY_CONSERVATION
            , vec3(max(cumulativeSpecularEnvironmentReflectance.r, max(cumulativeSpecularEnvironmentReflectance.g, cumulativeSpecularEnvironmentReflectance.b)))
        #else
            , baseSpecularEnvironmentReflectance
        #endif
```

Popov72 reviewed (commented):
Just a few more comments, hopefully the last ones!

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -123,17 +121,40 @@ reflectivityOutParams reflectivityBlock(
+            // This is an empirical hack to account for the fact that Schlick is tuned for IOR = 1.5
+            // and an IOR of 1.0 should result in no visible glancing specular.
+            float f90Scale = clamp(2.0 * (ior - 1.0), 0.0, 1.0);
+            outParams.reflectanceF90 = vec3(mix(specularWeight * f90Scale, 1.0, outParams.metallic));
```
Done. I also missed that F90 for metals was also affected by the F0Factor previously. This was the cause of the limb-darkening of metals in the furnace test image at the top of this PR.
I've reintroduced that now.

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockFinalLitComponents.fx:
```diff
@@ -6,24 +6,42 @@ aggShadow = aggShadow / numLights;
+                    baseSpecularEnergy *= baseSpecularEnergyConservationFactor;
+                #endif
+            #endif
+            finalIrradiance *= (vec3(1.0) - baseSpecularEnergy);
```
Done

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -98,43 +98,59 @@ reflectivityOutParams reflectivityBlock(
+        // and an IOR of 1.0 should result in no visible glancing specular.
+        float f90Scale = clamp(2.0 * (ior - 1.0), 0.0, 1.0);
+        #ifdef LEGACY_SPECULAR_ENERGY_CONSERVATION
+            outParams.reflectanceF90 = vec3(mix(specularWeight * f90Scale, specularWeight, outParams.metallic));
```
I think that this one should be:
```glsl
outParams.reflectanceF90 = vec3(specularWeight);
```
to be fully backward compatible?

MiiBond replied:
Done

Popov72 reviewed (commented):
Two small more comments after your changes.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -98,43 +98,59 @@ reflectivityOutParams reflectivityBlock(
+
+        #ifdef LEGACY_SPECULAR_ENERGY_CONSERVATION
+            surfaceReflectivityColor = mix(surfaceReflectivityColor, baseColor.rgb, outParams.metallic);
+            outParams.reflectanceF0 = max(surfaceReflectivityColor.r, max(surfaceReflectivityColor.g, surfaceReflectivityColor.b));
```
The code above sets `outParams.colorReflectanceF0 = mix(dielectricColorF0, metallicColorF0, outParams.metallic);` which, I think, should be equivalent to what was happening in the old code. Or did I miss something?

Popov72 replied:
My bad, the old code is doing `surfaceReflectivityColor = mix(metallicF0, baseColor, metallicRoughness.r);`, so all good!

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16476/merge/testResults/webgpuplaywright/index.html
